Attribute version: Added

mohit-0212 opened the pull request:
#62 Version Attribute is Added.

Verify for correctness. :)

achilleas-k commented on setup.py:
```diff
@@ -16,8 +16,9 @@
 if sys.version_info < (3, 4):
     install_req += ["enum"]
 
+exec(open('odml/version.py').read())
```
Using `exec` is generally a bad idea. There's no reason for that here, especially since you can just `from odml.version import __version__`.

mohit-0212 commented:
#68 Also resolved, along with simplified printing of xml output. Kindly review

achilleas-k commented:
So the new `version.py` file isn't needed anymore since it's been added to `addons.py`, right?

As for issue #68, I'll need to defer to @jgrewe for that one. Is that enough for what the issue was talking about?

Third note: Not sure why the build is failing. Maybe I was wrong to suggest importing the package into its own setup. I'll check it out and get back to you on that.

mohit-0212 commented:
Hey

I was about to get back to you on the same in a while. I was also not able
to get why the build was failing, and so was stuck on it for a while.
So I wanted to ask also how the testing is done there, so maybe I could
find my mistake.

1. Yes every thing added till now is in the addons file. I thought this
would be better instead of making different files for each, also later if
you want to remove any one of them you can do so.

2. I think so what I have done should be enough for the issue. If anything
else is to be added, you or @jgrewe can tell me, I'll try and add that
after.

3. As you said, I removed the exec command, and simply added from
odml.version import __version__ line. The required issue was fulfilled in
that case also, but couldn't figure why the build ended up failing.

Today, I'll try to work on the issue you suggested yesterday.

Also, kindly tell me what else should I do before starting writing my
proposal, so that I have a good chance.

Regards. Waiting for your reply.
Mohit


On Mar 9, 2017 4:29 PM, "Achilleas Koutsou" <notifications@github.com>
wrote:

So the new version.py file isn't needed anymore since it's been added to
addons.py, right?

As for issue #68 <https://github.com/G-Node/python-odml/issues/68>, I'll
need to defer to @jgrewe <https://github.com/jgrewe> for that one. Is that
enough for what the issue was talking about?

Third note: Not sure why the build is failing. Maybe I was wrong to suggest
importing the package into its own setup. I'll check it out and get back to
you on that.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/G-Node/python-odml/pull/76#issuecomment-285321182>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ATiMYXDSWC_oklqLaAywSVC4P1wucm5Mks5rj9uJgaJpZM4MWPq6>
.

achilleas-k commented:
Three comments:
1. Keep `version.py`. I understand it might seem messy to have tiny files for sitting around doing one small thing, but I prefer the purpose separation.
2. Read version in `setup.py` like it is done in [nixpy](https://github.com/G-Node/nixpy/blob/master/setup.py#L40). The reason for the error is because the setup script tries to import the library it's setting up, which is wrong and is totally my fault for suggesting it. Here's how I think it should be done:
    - Have `version.py` inside the `odml` package/directory.
    - Read in `version.py` like a normal file and parse the version string.
    - Assign the version string to the setup object.
    - Import `__version__` from `version.py` in `__init__.py` so it is available when importing the library as `odml.__version__`.
3. Backend selection in load/save convenience methods. The load and save methods in `addons.py` are nice, but they only use the XML parser. If we're going to add these, it would be better to have a way to select which backend to use, so maybe `load(filename)` could become `load(filename, backend="xml")`, such that it defaults to XML but the other readers can be used as well.
    - Related to this, the `print` function in `addons.py` is the old, Python2 style `print` which we don't use any more since the library is becoming Python 3 compatible.

mohit-0212 commented:
I corrected the version.py issue.
And also modified load, save, display for as required json support.
But I'm getting one error in load for json, so I have commented that part out for now.

The error when I run `odml.load(file,"json")` is the following:

`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mohit/Desktop/python-odml/odml/addons.py", line 8, in load
    return Y.JSONReader().fromFile(open(filename))
  File "/home/mohit/Desktop/python-odml/odml/tools/jsonparser.py", line 115, in fromFile
    return self.fromString(infile.read())
  File "/home/mohit/Desktop/python-odml/odml/tools/jsonparser.py", line 111, in fromString
    obj = json.loads(data)`

Till now I was able to trace back, but further, I couldn't

`File "/usr/lib/python3.4/json/__init__.py", line 318, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.4/json/decoder.py", line 343, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.4/json/decoder.py", line 361, in raw_decode
    raise ValueError(errmsg("Expecting value", s, err.value)) from None
ValueError: Expecting value: line 1 column 1 (char 0)`

Also, I don't know why the build failed for 1 check.

mohit-0212 commented:
I have above also resolved issue #81 .
I have used the SO link http://stackoverflow.com/questions/12605090/how-to-prevent-automatic-escaping-of-special-characters-in-python to solve this.

achilleas-k commented:
Since the version attribute was handled with PR #83, could you clean this PR up, rename it and rebase off current master?

I have a few comments on the other changes. I'll do a review on the individual lines for those.

achilleas-k commented on odml/addons.py:
```diff
@@ -0,0 +1,23 @@
+import odml.tools.xmlparser as X
+import odml.tools.jsonparser as Y 
+
```
Why import them as `X` and `Y`? Why not something meaningful like `xml` and `json`?

mohit-0212 replied:
Sure

achilleas-k commented on odml/addons.py:
```diff
@@ -0,0 +1,23 @@
+		X.XMLWriter(obj).write_file(path)
+	elif backend == "json":
+		Y.JSONWriter(obj).write_file(path)
+
```
Might be a good idea to throw an error if a bad backend is specified instead of failing silently.

mohit-0212 replied:
This doesn't fail, there is an error for JSONreader, which I mentioned. This works if we try to save a already loaded object.
Will nevertheless add a try-except, should have been there, as a fail-safe.

achilleas-k replied:
I mean if the method is called with `backend="foo"`, nothing will happen and no error will be thrown. The user will think the file was written, when it wasn't, and there will be no indication of the failure.

mohit-0212 replied:
These methods are shorter forms of the original methods, just to simplify the load/save. They just call the xmlparser/jsonparser inside odml/tools. The actual work done is by the methods inside them.
So if there exists an error, the method calls inside them which does the actual processing, would lead to the error, in my opinion.

jgrewe replied:
@mohit-0212 you are missing the point. 
Consider the user calls ```save(doc, 'somefile', backend="foo")```
What will happen? Nothing! The method will neither save the data nor raise an error. This is what is called a silent failure, there is no user feedback

mohit-0212 replied:
Okay, I was thinking about something else, totally my fault. Thanks for correcting :)

achilleas-k commented on odml/addons.py:
```diff
@@ -0,0 +1,23 @@
+		print (str(X.XMLWriter(obj)))
+	elif backend =="json":
+		print (str(Y.JSONWriter(obj)))
+
```
Same as above (failing silently).

achilleas-k commented on odml/doc.py:
```diff
@@ -25,10 +25,32 @@ class BaseDocument(base.sectionable, Document):
+                    new_value += char
+            return new_value
+        else:
+            return None
```
The `escape_ignore` method has a few issues.
1. Comparisons with `None` should be done using `is` and `is not` instead of `==` and `!=`.
2. The character-by-character iteration and appending is hugely inefficient. For one, it would be much cleaner to just use `str.replace()`, even though that's not great either.
3. Exception handling for each character compounds the problem even further. Escape characters are rare in the strings we're cleaning, so the `except` block will be run almost always while the `try` will run through very rarely. Having a non-escape character isn't _exceptional_.

Overall, though, are you sure this solves the issue in a general manner? Could you include some tests for this to check if we're getting the correct behaviour?

jgrewe commented on odml/addons.py:
```diff
@@ -0,0 +1,23 @@
+		
+
+def save(obj, path, backend="xml"):
+	if backend == "xml":
```
why is the interface for load and save different? In load you can specify a filename while save expects a path? What is the difference? I would rather have them named the same in both cases

jgrewe commented:
@mohit-0212 If you need help regarding rebasing squashing etc. let us know.

mohit-0212 commented:
@digaru19  Yes, just couldn't for past one day.

@jgrewe I tried rebasing, squashing; got some conflicts; will try to resolve them.
